### PR TITLE
Add Image Counter Lambda

### PR DESCRIPTION
**Pull request by stephengeller**
## What does this change?
This adds a lambda that gets the current image count from the media service API and posts it to CloudWatch, so that we can have monitoring and alerting on the image count.

## How can success be measured?
Whether, when deployed, we can execute the lambda and see the metric show up in CloudWatch.

## Who should look at this?

@guardian/digital-cms 

## Tested?
- [ ] locally
- [X] on TEST

### CODE REVIEW

File: image-counter-lambda-4621cc3c156f19193fcc [hunk 1]
```diff
@@ -0,0 +1 @@
+image-counter-lambda
```
Comment (akash1810) [inline image-counter-lambda-4621cc3c156f19193fcc]: This looks like an API key 😱

File: image-counter-lambda/src/handler.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+interface MediaAPICredentials {
+  baseUrl: string;
+  "X-Gu-Media-Key": string;
+}
+
+const getImageCount = async (
+  credentials: MediaAPICredentials
+): Promise<number> => {
+  const response = await fetch(credentials.baseUrl + "/images", {
```
Comment (akash1810) [inline image-counter-lambda/src/handler.ts]: We could use the `length` querystring to control the response size. That is, `length=0` will yield a smaller response than the default which is `length=10`.

File: scripts/ci/image-counter-lambda-build.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+apt-get update
+apt-get install zip
```
Comment (akash1810) [inline scripts/ci/image-counter-lambda-build.sh]: why is this needed?

File: image-counter-lambda-4621cc3c156f19193fcc (hunk 1 above)
Comment (stephengeller) [reply image-counter-lambda-4621cc3c156f19193fcc]: Sorry. It's been rotated out now.

File: image-counter-lambda/src/getCredentials.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+async function getCredentials() {
+  const s3 = new AWS.S3();
+  const params = {
+    Bucket: "grid-conf",
+    Key: `${process.env.STAGE}/image-counter-lambda/conf.json`
+  };
+
+  const data = await s3.getObject(params).promise();
+  const buffer = data.Body;
+  const body = buffer ? buffer.toString("utf8") : "";
```
Comment (jonathonherbert) [inline image-counter-lambda/src/getCredentials.ts]: NB: bonus new Typescript 3.7 feature -- in place of the ternary expression we should be able to write `buffer.toString?.("utf8") || ""` [thanks to optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) 😁

File: scripts/ci/image-counter-lambda-build.sh (hunk 3 above)
Comment (stephengeller) [reply scripts/ci/image-counter-lambda-build.sh]: We are running this CI script from inside a node docker image that doesn't have zip, and riffraff-artefact requires it. I personally think this is a bit cleaner than messing around with nvm stuff, as the image takes care of the node version by virtue of using `node:10`.

**Comment by stephengeller** [comment]
@akash1810 what did you have in mind for the logging? Not sure what you mean by strong structure, but keen to get some better logging in here for sure.

**Comment by akash1810** [comment]
> @akash1810 what did you have in mind for the logging? Not sure what you mean by strong structure, but keen to get some better logging in here for sure.

If we log a object, then we will create distinct fields in ELK. For example:

```js
const logDetail = {
  count: 100
};

console.log("retrieved count", logDetail)
```

This will create a `count` field in ELK which we can use to graph on.

File: media-api/conf/routes [hunk 5]
```diff
@@ -35,6 +35,7 @@ GET     /usage/quotas/:id                             controllers.UsageControlle
 # Management
 GET     /management/healthcheck                       com.gu.mediaservice.lib.management.ElasticSearchHealthCheck.healthCheck
 GET     /management/manifest                          com.gu.mediaservice.lib.management.ManagementWithPermissions.manifest
+GET     /management/imageCounts                       com.gu.mediaservice.lib.management.ElasticSearchHealthCheck.imageCounts
```
Comment (akash1810) [inline media-api/conf/routes]: Could we split the additional route into another PR? This is because the order of deployment isn't completely deterministic; if the lambda was to ship first it'll get a 404 until media-api had deployed too.

Comment (stephengeller) [reply media-api/conf/routes]: [This has been moved here](https://github.com/guardian/grid/pull/2735)

File: common-lib/src/main/scala/com/gu/mediaservice/lib/management/Management.scala [hunk 6]
```diff
@@ -63,4 +66,20 @@ class ElasticSearchHealthCheck(override val controllerComponents: ControllerComp
       }
     }
   }
+
+  def imageCounts: Action[AnyContent] = Action.async {
+    elasticsearch.countImages().map {
+      case _ @(catCount: CatCountResponse, searchResponse: SearchResponse, indexStats: IndexStatsResponse) =>
+        val json: String =
+          s"""{
+             |  "catCount": ${catCount.count},
```
[the remark this answers is not in the document]
Comment (stephengeller) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/management/Management.scala]: > > @akash1810 what did you have in mind for the logging? Not sure what you mean by strong structure, but keen to get some better logging in here for sure.
> 
> If we log a object, then we will create distinct fields in ELK. For example:
> 
> ```js
> const logDetail = {
>   count: 100
> };
> 
> console.log("retrieved count", logDetail)
> ```
> 
> This will create a `count` field in ELK which we can use to graph on.

This has now been added, take a look at the lambda function on line 48.

File: scripts/ci/image-counter-lambda-build.sh (hunk 3 above)
Comment (AWare) [reply scripts/ci/image-counter-lambda-build.sh]: I just saw this, and this is fixed by using node-riffraff-artifact
